src/nodekit.ts

```diff
@@ -91,25 +91,33 @@ export class NodeKit {
             utils: this.utils,
         });
 
-        this.addShutdownHandler(() => {
-            return new Promise((resolve) => {
-                this.tracer.close?.(() => resolve());
-            });
-        });
+        this.addShutdownHandler(() => new Promise<void>((resolve) => this.tracer.close(resolve)));
 
-        ['SIGTERM', 'SIGINT'].forEach((signal) => {
-            process.on(signal, () => {
-                Promise.all(this.shutdownHandlers.map((handler) => handler()))
-                    .then(() => process.exit(0))
-                    .catch((error) => {
-                        this.ctx.logError('Error executing shutdown handlers', error);
-                        process.exit(1);
-                    });
-            });
-        });
+        this.setupShutdownSignals();
     }
 
     addShutdownHandler(handler: ShutdownHandler) {
         this.shutdownHandlers.push(handler);
     }
+
+    private setupShutdownSignals() {
+        const signals = ['SIGTERM', 'SIGINT'] as const;
+
+        const handleSignal: ShutdownHandler = (signal) => {
+            signals.forEach((off) => process.off(off, handleSignal));
+
+            let code = 0;
+
+            const promises = this.shutdownHandlers.map((handler) =>
+                new Promise((resolve) => resolve(handler(signal))).catch((error) => {
+                    code = 1;
+                    this.ctx.logError('Error executing shutdown handler', error);
+                }),
+            );
+
+            Promise.allSettled(promises).then(() => process.exit(code));
+        };
+
+        signals.forEach((signal) => process.on(signal, handleSignal));
```

Comment:
Does it pass signal name to the handler?

Reply:
It is not properly typed but yes, signal name would be passed.

https://nodejs.org/api/process.html#signal-events